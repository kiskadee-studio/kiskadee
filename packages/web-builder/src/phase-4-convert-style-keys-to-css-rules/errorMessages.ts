export const UNSUPPORTED_PROPERTY_NAME = (propertyName: string, styleKey: string): string =>
  `Invalid style key "${styleKey}". The property name must be "${propertyName}".`;

export const UNSUPPORTED_VALUE = (
  propertyName: string,
  invalidValue: string,
  styleKey: string
): string =>
  `Unsupported value "${invalidValue}" for property "${propertyName}" in style key "${styleKey}".`;

// Shadow ------------------------------------------------------------------------------------------

export const UNSUPPORTED_INTERACTION_STATE = (state: string, styleKey: string): string =>
  `Unsupported interaction state "${state}" in shadow key "${styleKey}".`;

export const INVALID_SHADOW_COLOR_VALUE = 'Invalid shadow color value.';
