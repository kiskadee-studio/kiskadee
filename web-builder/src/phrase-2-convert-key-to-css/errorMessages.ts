export const UNSUPPORTED_PROPERTY = (expectedProperty: string, styleKey: string): string =>
  `Invalid style key "${styleKey}". Expected the key to be in the format "<property>__<value>" with the property part being "${expectedProperty}".`;

export const UNSUPPORTED_VALUE = (
  property: string,
  receivedValue: string,
  styleKey: string
): string =>
  `Unsupported value "${receivedValue}" for property "${property}" in style key "${styleKey}".`;
