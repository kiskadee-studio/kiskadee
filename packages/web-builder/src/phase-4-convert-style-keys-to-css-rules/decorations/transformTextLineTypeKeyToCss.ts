import {
  CssDecorationProperty,
  CssTextDecorationValue,
  type DecorationProperty,
  type TextLineTypeValue
} from '@kiskadee/schema';
import { UNSUPPORTED_PROPERTY_NAME, UNSUPPORTED_VALUE } from '../errorMessages';
import type { GeneratedCss } from '../phrase2.types';

const { textLineType }: typeof CssDecorationProperty = CssDecorationProperty;

/**
 * Converts a text‐decoration style key into a GeneratedCss object.
 *
 * The key must be in the format "textDecoration__<value>", where <value> is one of the
 * values defined in CssTextDecorationValue.
 *
 * @example
 * ```ts
 * const result = transformTextDecorationKeyToCss("textDecoration__underline");
 * // result:
 * // {
 * //   className: "textDecoration__underline",
 * //   cssRule: ".textDecoration__underline { text-decoration: underline; }"
 * // }
 * ```
 *
 * @param styleKey - The text‐decoration key (e.g. "textDecoration__underline")
 * @returns A GeneratedCss object containing:
 *   - `className`: the raw key (no leading dot)
 *   - `cssRule`: the full CSS rule string
 * @throws Error if the property isn’t "textDecoration", the format is wrong or the value is unsupported.
 */
export function transformTextLineTypeKeyToCss(styleKey: string): GeneratedCss {
  const propertyName: DecorationProperty = 'textLineType';
  const styleParts = styleKey.split('__');
  const styleProperty = styleParts[0];
  const styleValue = styleParts[1] as TextLineTypeValue;

  if (styleProperty !== propertyName) {
    throw new Error(UNSUPPORTED_PROPERTY_NAME(propertyName, styleKey));
  }

  if (styleParts.length !== 2) {
    const [, ...invalidValue] = styleParts;
    throw new Error(UNSUPPORTED_VALUE(propertyName, invalidValue.join('__'), styleKey));
  }

  const value: CssTextDecorationValue | undefined = CssTextDecorationValue[styleValue];

  if (value === undefined) {
    throw new Error(UNSUPPORTED_VALUE(propertyName, styleValue, styleKey));
  }

  const cssRule = `.${styleKey} { ${textLineType}: ${value}; }`;

  return {
    className: styleKey,
    cssRule
  };
}
