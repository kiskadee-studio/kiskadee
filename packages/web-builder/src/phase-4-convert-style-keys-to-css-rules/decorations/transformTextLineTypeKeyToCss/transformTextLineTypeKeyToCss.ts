import {
  CssDecorationProperty,
  CssTextDecorationValue,
  type DecorationProperty,
  type TextLineTypeValue
} from '@kiskadee/schema';
import { UNSUPPORTED_PROPERTY_NAME, UNSUPPORTED_VALUE } from '../../errorMessages';

const { textLineType }: typeof CssDecorationProperty = CssDecorationProperty;

/**
 * Converts a text‐decoration style key into a CSS rule string.
 *
 * The key must be in the format "textDecoration__<value>", where <value> is one of the
 * values defined in CssTextDecorationValue.
 *
 * @example
 * ```ts
 * const css = transformTextLineTypeKeyToCss("textLineType__underline", "textLineType__underline");
 * // css: ".textLineType__underline { text-decoration: underline; }"
 * ```
 *
 * @param styleKey - The text‐decoration key (e.g. "textLineType__underline")
 * @param className - The class name to use in the generated CSS (without leading dot).
 *                    Usually this will be the same as styleKey or an optimized short name.
 * @returns The generated CSS rule string (selector + declarations).
 * @throws Error if the property isn’t "textLineType", the format is wrong or the value is unsupported.
 */
export function transformTextLineTypeKeyToCss(styleKey: string, className: string): string {
  const propertyName: DecorationProperty = 'textLineType';
  const styleParts = styleKey.split('__');
  const styleProperty = styleParts[0];
  const styleValue = styleParts[1] as TextLineTypeValue;

  if (styleProperty !== propertyName) {
    throw new Error(UNSUPPORTED_PROPERTY_NAME(propertyName, styleKey));
  }

  if (styleParts.length !== 2) {
    const [, ...invalidValue] = styleParts;
    throw new Error(UNSUPPORTED_VALUE(propertyName, invalidValue.join('__'), styleKey));
  }

  const value: CssTextDecorationValue | undefined = CssTextDecorationValue[styleValue];

  if (value === undefined) {
    throw new Error(UNSUPPORTED_VALUE(propertyName, styleValue, styleKey));
  }

  return `.${className} { ${textLineType}: ${value} }`;
}
