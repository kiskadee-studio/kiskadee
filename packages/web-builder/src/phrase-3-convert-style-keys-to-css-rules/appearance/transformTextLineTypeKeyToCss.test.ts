import { describe, it, expect } from 'vitest';
import {
  CssDecorationProperty,
  CssTextDecorationValue,
  type TextLineTypeProperty,
  type TextLineTypeValue
} from '@kiskadee/schema';
import { transformTextLineTypeKeyToCss } from './transformTextLineTypeKeyToCss';
import { UNSUPPORTED_PROPERTY, UNSUPPORTED_VALUE } from '../errorMessages';

const propertyName: TextLineTypeProperty = 'textLineType';
const { textLineType } = CssDecorationProperty;
const { lineThrough, underline, none } = CssTextDecorationValue;

describe('transformTextLineTypeKeyToCss', () => {
  describe('Successful operation', () => {
    it('should return CSS for text line type "underline"', () => {
      const textLineTypeValue: TextLineTypeValue = 'underline';
      const styleKey = `${propertyName}__${textLineTypeValue}`;
      const expectedRule = `.${styleKey} { ${textLineType}: ${underline}; }`;
      const result = transformTextLineTypeKeyToCss(styleKey);

      expect(result).toEqual({
        className: styleKey,
        cssRule: expectedRule
      });
      expect(result).toMatchSnapshot();
    });

    it('should return CSS for text line type "lineThrough"', () => {
      const textLineTypeValue: TextLineTypeValue = 'lineThrough';
      const styleKey = `${propertyName}__${textLineTypeValue}`;
      const expectedRule = `.${styleKey} { ${textLineType}: ${lineThrough}; }`;
      const result = transformTextLineTypeKeyToCss(styleKey);

      expect(result).toEqual({
        className: styleKey,
        cssRule: expectedRule
      });
      expect(result).toMatchSnapshot();
    });

    it('should return CSS for text line type "none"', () => {
      const textLineTypeValue: TextLineTypeValue = 'none';
      const styleKey = `${propertyName}__${textLineTypeValue}`;
      const expectedRule = `.${styleKey} { ${textLineType}: ${none}; }`;
      const result = transformTextLineTypeKeyToCss(styleKey);

      expect(result).toEqual({
        className: styleKey,
        cssRule: expectedRule
      });
      expect(result).toMatchSnapshot();
    });
  });

  describe('Error handling', () => {
    describe('Unsupported property errors', () => {
      it('throws error if style key prefix is incorrect', () => {
        const invalidProperty = 'invalidProperty';
        const styleKey = `${invalidProperty}__underline`;
        const expectedMessage = UNSUPPORTED_PROPERTY(propertyName, styleKey);

        let caught: unknown;
        try {
          transformTextLineTypeKeyToCss(styleKey);
        } catch (err) {
          caught = err;
        }

        expect(caught).toBeInstanceOf(Error);
        const error = caught as Error;
        expect(error.message).toBe(expectedMessage);
        expect(error.message).toMatchSnapshot();
      });

      it('throws error if style key format is invalid', () => {
        const styleKey = 'textLineType-underline';
        const expectedMessage = UNSUPPORTED_PROPERTY(propertyName, styleKey);

        let caught: unknown;
        try {
          transformTextLineTypeKeyToCss(styleKey);
        } catch (err) {
          caught = err;
        }

        expect(caught).toBeInstanceOf(Error);
        const error = caught as Error;
        expect(error.message).toBe(expectedMessage);
        expect(error.message).toMatchSnapshot();
      });
    });

    describe('Unsupported value errors', () => {
      it('throws error if the value part is missing', () => {
        const styleKey = `${propertyName}__`;
        const expectedMessage = UNSUPPORTED_VALUE(propertyName, '', styleKey);

        let caught: unknown;
        try {
          transformTextLineTypeKeyToCss(styleKey);
        } catch (err) {
          caught = err;
        }

        expect(caught).toBeInstanceOf(Error);
        const error = caught as Error;
        expect(error.message).toBe(expectedMessage);
        expect(error.message).toMatchSnapshot();
      });

      it('throws error if value contains unexpected separators', () => {
        const invalidValue = 'underline__dotted';
        const styleKey = `${propertyName}__${invalidValue}`;
        const expectedMessage = UNSUPPORTED_VALUE(propertyName, invalidValue, styleKey);

        let caught: unknown;
        try {
          transformTextLineTypeKeyToCss(styleKey);
        } catch (err) {
          caught = err;
        }

        expect(caught).toBeInstanceOf(Error);
        const error = caught as Error;
        expect(error.message).toBe(expectedMessage);
        expect(error.message).toMatchSnapshot();
      });

      it('throws error if value is unsupported', () => {
        const invalidValue = 'overline';
        const styleKey = `${propertyName}__${invalidValue}`;
        const expectedMessage = UNSUPPORTED_VALUE(propertyName, invalidValue, styleKey);

        let caught: unknown;
        try {
          transformTextLineTypeKeyToCss(styleKey);
        } catch (err) {
          caught = err;
        }

        expect(caught).toBeInstanceOf(Error);
        const error = caught as Error;
        expect(error.message).toBe(expectedMessage);
        expect(error.message).toMatchSnapshot();
      });
    });
  });
});
