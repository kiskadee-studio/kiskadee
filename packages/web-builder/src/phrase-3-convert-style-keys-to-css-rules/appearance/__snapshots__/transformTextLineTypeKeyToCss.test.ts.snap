// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported property errors > should throw when style key does not start with "textLineType__" 1`] = `"Invalid style key "invalidProperty__underline". Expected the key to be in the format "<property>__<value>" with the property part being "textLineType"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported property errors > should throw when the key cannot be split correctly 1`] = `"Invalid style key "textLineType-underline". Expected the key to be in the format "<property>__<value>" with the property part being "textLineType"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported property errors > throws error if style key format is invalid 1`] = `"Invalid style key "textLineType-underline". Expected the key to be in the format "<property>__<value>" with the property part being "textLineType"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported property errors > throws error if style key prefix is incorrect 1`] = `"Invalid style key "invalidProperty__underline". Expected the key to be in the format "<property>__<value>" with the property part being "textLineType"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > should throw when no text line type value is provided 1`] = `"Unsupported value "" for property "textLineType" in style key "textLineType__"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > should throw when style key contains extra separators 1`] = `"Unsupported value "underline__dotted" for property "textLineType" in style key "textLineType__underline__dotted"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > should throw when text line type value is unsupported 1`] = `"Unsupported value "overline" for property "textLineType" in style key "textLineType__overline"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > throws error if the value part is missing 1`] = `"Unsupported value "" for property "textLineType" in style key "textLineType__"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > throws error if value contains unexpected separators 1`] = `"Unsupported value "underline__dotted" for property "textLineType" in style key "textLineType__underline__dotted"."`;

exports[`transformTextLineTypeKeyToCss > Error handling > Unsupported value errors > throws error if value is unsupported 1`] = `"Unsupported value "overline" for property "textLineType" in style key "textLineType__overline"."`;

exports[`transformTextLineTypeKeyToCss > Successful operation > should return CSS for text line type "lineThrough" 1`] = `
{
  "className": "textLineType__lineThrough",
  "cssRule": ".textLineType__lineThrough { text-decoration: line-through; }",
}
`;

exports[`transformTextLineTypeKeyToCss > Successful operation > should return CSS for text line type "none" 1`] = `
{
  "className": "textLineType__none",
  "cssRule": ".textLineType__none { text-decoration: none; }",
}
`;

exports[`transformTextLineTypeKeyToCss > Successful operation > should return CSS for text line type "underline" 1`] = `
{
  "className": "textLineType__underline",
  "cssRule": ".textLineType__underline { text-decoration: underline; }",
}
`;
